x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff_outcome.ts

```diff
@@ -70,3 +118,29 @@ export const determineIfValueCanUpdate = (diffCase: ThreeWayDiffOutcome): boolea
     diffCase === ThreeWayDiffOutcome.CustomizedValueCanUpdate
   );
 };
+
+/**
+ * Returns a boolean if 2 arrays contain same the elements agnostic of order after being deduplicated
+ *
+ * NOTE: array equality is case insensitive
+ */
+const arraysHaveSameElements = <T>(arr1: T[], arr2: T[]) => {
+  const set1 = new Set(
+    arr1.map((value) => (typeof value === 'string' ? value.toLowerCase() : value))
+  );
+  const set2 = new Set(
+    arr2.map((value) => (typeof value === 'string' ? value.toLowerCase() : value))
+  );
```

Comment:
I agree with deduplication, because for our most used rule fields duplication doesn't change any semantics:

- `author`: `['Elastic']` and `['Elastic', 'Elastic']` are semantically the same values
- `tags`: `['Windows']` and `['Windows', 'Windows']` are semantically the same values, and two rules with such tags will work the same way in terms of filtering by tags
- `index`: `['logs-*']` and `['logs-*', 'logs-*']` are semantically the same values, and two rules with such index patterns will work the same way in terms of querying source indices

However, I disagree with making this algorithm case-insensitive for string arrays. Most (all?) of our rule fields are case sensitive:

- Alerting Framework's tags are case sensitive
- Elasticsearch indices MUST be lowercase, which means it's important for index patterns to be lowercase as well => it will be important to have a case-sensitive algorithm for index patterns that can reveal typos, such as `logs-*` vs `Logs-*`.
- Authors are people or company names, obviously these are case sensitive.
- Licenses
- etc

What was your reasoning for introducing case sensitivity?

Reply:
Hi @banderror thanks for this important catch.

The original thinking here was that the app currently tries to "steer" the user away from duplicating tags or arrays which are the same but with different case - at least in the UI. For example, if a rule has an index pattern of `endgame-*`, the UI blocks you from adding `ENDGAME-*` or `Endgame-*`. So the idea was to follow that behaviour and consider `endgame-*` and `ENDGAME-*` to be "the same" when diffing.

But yes, I see now that would actually lead to unexpected behaviour for the rule: if we consider those two to be the same during diffing, and the rule actually updates to `ENDGAME-*`, the rule won't work as expected. And same applies to tags.